impl/src/settings.rs

```diff
@@ -367,8 +380,95 @@ pub fn format_registry_url(registry_url: &str, name: &str, version: &str) -> Str
     .replace("{version}", version)
 }
 
+fn validate_crate_setting_additional_build_file_is_relative(
+  file_path: &Path,
+  crate_name: &str,
+  version: &VersionReq,
+) -> Result<(), RazeError> {
+  if file_path.is_absolute() {
+    return Err(RazeError::Config {
+      field_path_opt: Some(format!(
+        "raze.crates.{}.{}.additional_build_file",
+        crate_name,
+        version.to_string()
+      )),
+      message: "additional_build_file flags should be reletive paths from the manifest containing \
+                the definition"
+        .to_owned(),
+    });
+  }
+
+  Ok(())
+}
+
+/** Ensures crate settings which specify additional build files have canonicalized paths */
+fn validate_crate_setting_additional_build_file(
+  additional_build_file: &mut &mut PathBuf,
```

Comment:
Hey, what's going on here?

Are you using the validation function to canonicalize this path? This both syntactically is weird, and also represents a really weird and unintuitive thing for the settings code to do which I think will be confusing to future readers. Is it really not possible to do this work during catalog construction? I see you only need the toml. Are we losing the toml or something?

Reply:
So is this being handled here: https://github.com/google/cargo-raze/pull/301#discussion_r539820850 ?

Reply:
No. The opposite. Fixing this problem fixes that problem.

Reply:
This is where that field gets canonicalized. How can I go about doing that?

Reply:
Also, this method currently mutates `RazeSettings` https://github.com/google/cargo-raze/blob/master/impl/src/settings.rs#L387

Reply:
Can you suggest a better pattern here?

Reply:
Actually that whole function does some sanitation next to validation. Would I be able to address the mixup there in another PR?

Reply:
Still happy to make some improvements here. Just don't want to caught up doing code cleanup (in this PR).

Reply:
I wrote this comment two times already and github ate my comment both times (once because I opened the code to look at the context, and once because I accidentally ctrl-r). I am grump. This version is going to be more terse than the last two times.

So, yes, that code's existence means that you can't remove the &mut below. It's also funny, because I wrote that code (ouch).

There's a difference here. My code standardizes the path so that future code can operate on it without dealing with the presence or absence of the slash. A developer who specifies a path while debugging something is going to see approximately the same path in their toml, and in their code and will not be confused. Yours, on the other hand, is making a significant and meaningful change to the setting value of the same name. What you see in the toml, and what you see in the planning code for the value of this field is not the same. This is an nuanced argument, but I think it holds up.

In my code, I could've been consistent in this principle by either (1) telling users not to end their paths in a slash (user hostile!), or (2) updating the code which accesses this field to support *either* a trailing slash *or* no slash.

---

I'd like to provide a better suggestion, but I don't understand how this PR, and #302 are parsing the individual toml files. My initial suggestion was to retain the cargo toml path, and canonicalize using that path during render_planned_build (where you're accessing the additional_build_path). I recognized that this would only work for the root toml though, so I referenced #302 to understand how you're properly canonicalizing settings from workspace subcrates, and I don't think you are. validate_settings's toml path comes from load_settings. In #302, load_settings is only being called once, with one toml path, which is the root toml path.

Am I missing something here?

Reply:
My working understanding is that canonicalization requires access to the workspace subcrate's toml. My lack of clarity above is basically that your next PR is not correctly handling this anyway. If your next PR *is* correct, and in fact, you *can* do canonicalization with only the workspace root toml path, then I don't see what the issue at all is. Canonicalize in render_planned_build using the path to the workspace root toml.

Reply:
I might be wrong on this. My initial analysis was based on some ctrl+f-ing, and it seems i didn't expand settings.rs

Reply:
No, I think I'm right. 

You're parsing out the settings from the workspace crates in parse_raze_settings_workspace. You're validating these settings all together with the concatenated settings in validate_settings (after calling `parse_raze_settings`) only once, only with the root toml.

Reply:
The future I'm trying to get to is that all raze settings are parsed via cargo metadata. There are two places for metadata, packages and the workspace itself. When we're parsing package metadata, (specifically the `additional_build_file` argument), it will be a relative path to a file in the repo (or it historically has been). If we parse a package's metadata, we'll have that file path as it's written. This is not sufficient in future phases because it's a relative path to a file *from* the `Cargo.toml` file it's been defined in. Because all crate settings are stored in a HashMap `pub crates: HashMap<String, CrateSettingsPerVersion>,` (and because we don't track the path to the package a setting came from), we are unable to find the actually file on disk.

Reply:
Oh, I think I see what you're saying. This might be a bug

Reply:
Yeah, this was definitely not doing what I had intended. I've updated it. The path is parsed as-is for `RazeSettings` and the path is updated during the `planning` phase where it's canonicalized. Paths must be relative paths from the *Cargo* workspace root (which should match the current behavior). It'll be a bit awkward when defining that field in multiple packages but I think it'll be fine and perhaps this can be improved in the future (or maybe it wont' really be a problem, time will tell).